modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticMohrCoulombMulti.C

```diff
@@ -252,173 +351,547 @@ TensorMechanicsPlasticMohrCoulombMulti::activeConstraints(const std::vector<Real
+  if (yf[1] > _f_tol && yf[3] > _f_tol && yf[5] > _f_tol)
   {
-    act[0] = act[1] = act[3] = true;
-    returned_stress(0, 0) = returned_stress(1, 1) = returned_stress(2, 2) = cohcot;
-    returned_stress = eigvecs*returned_stress*(eigvecs.transpose());
-    return;
+    trial_order[0] = tip110100;
+    trial_order[1] = edge010100;
+    trial_order[2] = plane000100;
+    trial_order[3] = edge000101;
+    trial_order[4] = tip010101;
   }
-
-  // 010101 (tip)
-  denom = triple(n[1], n[3], n[5]);
-  if (triple(v, n[1], n[3])/denom >= 0 && triple(v, n[3], n[5])/denom >= 0 && triple(v, n[5], n[1])/denom >= 0)
+  else if (yf[1] <= _f_tol && yf[3] > _f_tol && yf[5] > _f_tol)
   {
-    act[1] = act[3] = act[5] = true;
-    returned_stress(0, 0) = returned_stress(1, 1) = returned_stress(2, 2) = cohcot;
-    returned_stress = eigvecs*returned_stress*(eigvecs.transpose());
-    return;
+    trial_order[0] = edge000101;
+    trial_order[1] = plane000100;
+    trial_order[2] = tip110100;
+    trial_order[3] = tip010101;
+    trial_order[4] = edge010100;
+  }
+  else if (yf[1] <= _f_tol && yf[3] > _f_tol && yf[5] <= _f_tol)
+  {
+    trial_order[0] = plane000100;
+    trial_order[1] = edge000101;
+    trial_order[2] = edge010100;
+    trial_order[3] = tip110100;
+    trial_order[4] = tip010101;
+  }
+  else
+  {
+    trial_order[0] = edge010100;
+    trial_order[1] = plane000100;
+    trial_order[2] = edge000101;
+    trial_order[3] = tip110100;
+    trial_order[4] = tip010101;
   }
 
+  unsigned trial;
+  bool nr_converged;
+  bool kt_success;
+  std::vector<std::vector<Real> > ntip(3);
+  ntip[0].resize(3);
+  ntip[1].resize(3);
+  ntip[2].resize(3);
+  std::vector<Real> dpmtip(3);
 
-  // the following are tangents to the 1, 3, and 5 yield surfaces
-  // used below.
-  std::vector<Real> n1xn3(3);
-  cross(n[1], n[3], n1xn3);
-  std::vector<Real> n1xn5(3);
-  cross(n[1], n[5], n1xn5);
-  std::vector<Real> n3xn5(3);
-  cross(n[3], n[5], n3xn5);
-
-
-
-  // 010100 (edge)
-  // work out the point to which we would return, "a".  It is defined by
-  // f1 = 0 = f3, and that (p - a).(n1 x n3) = 0, where "p" is the
-  // starting position (p = eigvals).
-  // In the following a = (lam0, lam1, lam1)
-  Real pdotn1xn3 = dot(eigvals, n1xn3);
-  //Real lam0 = pdotn1xn3 - 4*coh*cosphi*oneminus/(1 + sinphi);
-  Real lam0 = pdotn1xn3 - 2*coh*cosphi*(n1xn3[1] + n1xn3[2])/(1 + sinphi);
-  //lam0 /= oneplus + 2*oneminus*(1 - sinphi)/(1 + sinphi);
-  lam0 /= n1xn3[0] + (n1xn3[1] + n1xn3[2])*(1 - sinphi)/(1 + sinphi);
-  Real lam1 = lam0*(1 - sinphi)/(1 + sinphi) + 2*coh*cosphi/(1 + sinphi);
-  std::vector<Real> pminusa(3);
-  pminusa[0] = eigvals[0] - lam0;
-  pminusa[1] = eigvals[1] - lam1;
-  pminusa[2] = eigvals[2] - lam1;
-  // p = a + alpha*n1 + beta*n3, and for Kuhn-Tucker we need alpha and beta non-negative
-  if (dot(pminusa, n3xn5)/dot(n[1], n3xn5) >= 0 && dot(pminusa, n1xn5)/dot(n[3], n1xn5) >= 0)
+  for (trial = 0 ; trial < 5 ; ++trial)
```

Comment:
And what is the significance of `5` trials? Can this be a `const unsigned int` with a descriptive name, because then...